File: scripts/translate-fr-to-en.mjs
import fs from 'node:fs'
import path from 'node:path'
import yaml from 'js-yaml'

const baseDirs = [
  path.join('src', 'components'),
  path.join('src', 'stores'),
  path.join('src', 'data'),
]

async function translate(text) {
  const params = new URLSearchParams({ q: text, langpair: 'fr|en' })
  const res = await fetch('https://api.mymemory.translated.net/get', {
    method: 'POST',
    headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
    body: params,
  })
  const json = await res.json()
  return json.responseData.translatedText
}

async function processFile(file) {
  const content = fs.readFileSync(file, 'utf8')
  const data = yaml.load(content)
  const root = Object.keys(data)[0]
  const fr = data[root].fr
  if (!fr || data[root].en)
    return
  const en = {}
  for (const key of Object.keys(fr)) {
    if (typeof fr[key] === 'string')
      en[key] = await translate(fr[key])
  }
  data[root].en = en
  fs.writeFileSync(file, yaml.dump(data, { lineWidth: 0 }))
  console.log('Updated', file)
}

async function main() {
  for (const base of baseDirs) {
    if (!fs.existsSync(base))
      continue
    const files = []
    function walk(dir) {
      for (const entry of fs.readdirSync(dir, { withFileTypes: true })) {
        const res = path.join(dir, entry.name)
        if (entry.isDirectory())
          walk(res)
        else if (entry.isFile() && res.endsWith('.i18n.yml'))
          files.push(res)
      }
    }
    walk(base)
    for (const f of files)
      await processFile(f)
  }
}

main()

File: scripts/translate-fr-to-en.py
#!/usr/bin/env python3
import os
import yaml
import argostranslate.package
import argostranslate.translate

langs = argostranslate.translate.get_installed_languages()
fr = next(l for l in langs if l.code == 'fr')
en = next(l for l in langs if l.code == 'en')
translate = fr.get_translation(en).translate

base_dirs = ['src/components', 'src/stores', 'src/data']

for base in base_dirs:
    if not os.path.isdir(base):
        continue
    for root, _, files in os.walk(base):
        for file in files:
            if not file.endswith('.i18n.yml'):
                continue
            path = os.path.join(root, file)
            data = yaml.safe_load(open(path))
            key = next(iter(data))
            fr_data = data[key].get('fr')
            if not fr_data or data[key].get('en'):
                continue
            en_trans = {}
            for k, v in fr_data.items():
                if isinstance(v, str):
                    en_trans[k] = translate(v)
            data[key]['en'] = en_trans
            with open(path, 'w') as f:
                yaml.dump(data, f, allow_unicode=True, width=float('inf'))
            print('Updated', path)

